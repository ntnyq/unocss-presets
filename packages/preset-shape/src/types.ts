export interface PresetShapeOptions {
  /**
   * customize shape selector prefix
   *
   * @default 'shape-'
   */
  prefix?: string
}
