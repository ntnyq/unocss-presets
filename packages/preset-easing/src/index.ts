import { camelCase, kebabCase } from '@ntnyq/utils'
import { definePreset } from '@unocss/core'
import { easings } from './easings'
import type { PresetEasingOptions } from './types'

export const presetEasing = definePreset<PresetEasingOptions>((options = {}) => {
  const { prefix = 'easing-' } = options
  const selectorPattern = `${prefix}(${Object.keys(easings)
    .map(name => kebabCase(name))
    .join('|')})`

  return {
    name: 'unocss-preset-easing',

    rules: [
      [
        new RegExp(`^${selectorPattern}$`),
        ([, name]) => ({
          'transition-timing-function': easings[camelCase(name) as keyof typeof easings],
        }),
        {
          autocomplete: [selectorPattern],
        },
      ],
    ],
  }
})

export * from './types'
export * from './easings'

export default presetEasing
