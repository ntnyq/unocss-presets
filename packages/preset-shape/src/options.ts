import type { PresetShapeOptions } from './types'

export const DEFAULT_OPTIONS: Required<PresetShapeOptions> = {
  prefix: 'shape-',
}

export function resolveOptions(
  options: PresetShapeOptions = {},
): Required<PresetShapeOptions> {
  return {
    ...DEFAULT_OPTIONS,
    ...options,
  }
}
