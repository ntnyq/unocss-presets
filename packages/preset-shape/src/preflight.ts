import { createCSSVar } from '../utils'
import type { Preflight } from '@unocss/core'
import type { PresetShapeOptions } from './types'

export const createPreflights = (options: Required<PresetShapeOptions>): Preflight[] => {
  const cssVar = createCSSVar(options.prefix)
  const cssVarUse = createCSSVar(options.prefix, { use: true })
  return [
    {
      getCSS: () =>
        `:root {
${[
  `${cssVar('star-clip-path')}: polygon(50% 0,79% 90%,2% 35%,98% 35%,21% 90%);`,
  `${cssVar('triangle-clip-path')}: polygon(50% 0,100% 100%,0 100%);`,
  `${cssVar('rhombus-clip-path')}: polygon(50% 0,100% 50%,50% 100%,0 50%);`,
  `${cssVar('hexagon-clip-path')}: polygon(50% -50%,100% 50%,50% 150%,0 50%);`,

  `${cssVar('octagon-angle')}: calc(50% * tan(-22.5deg));`,
  `${cssVar('octagon-clip-path')}: polygon(${cssVarUse('octagon-angle')} 50%, 50% ${cssVarUse('octagon-angle')}, calc(100% - ${cssVarUse('octagon-angle')}) 50%, 50% calc(100% - ${cssVarUse('octagon-angle')}));`,

  `${cssVar('heart-gradient-mask')}: radial-gradient(#000 69%,#0000 70%) 84.5%/50%;`,
  `${cssVar('heart-clip-path')}: polygon(-41% 0,50% 91%, 141% 0);`,

  `${cssVar('parallelogram-skew')}: 30%;`,
  `${cssVar('parallelogram-clip-path')}: polygon(${cssVarUse('parallelogram-skew')} 0,100% 0,calc(100% - ${cssVarUse('parallelogram-skew')}) 100%,0 100%);`,
].join('\n')}
}`.trim(),
    },
  ]
}
