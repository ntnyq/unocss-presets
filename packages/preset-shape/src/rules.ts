import { createCSSVar } from '../utils'
import type { Rule } from '@unocss/core'
import type { PresetShapeOptions } from './types'

export const createShapeRules = (options: Required<PresetShapeOptions>) => {
  const cssVarUse = createCSSVar(options.prefix, { use: true })
  const rules: Rule[] = [
    [
      new RegExp(`^${options.prefix}star-?(.+)$`),
      ([_, s]) => ({
        width: s,
        'aspect-ratio': 1,
        'clip-path': cssVarUse('star-clip-path'),
      }),
      {
        autocomplete: `${options.prefix}star-$size`,
      },
    ],

    [
      new RegExp(`^${options.prefix}triangle-?(.+)$`),
      ([_, s]) => ({
        width: s,
        'aspect-ratio': '1/cos(30deg)',
        'clip-path': cssVarUse('triangle-clip-path'),
      }),
      {
        autocomplete: `${options.prefix}triangle-$size`,
      },
    ],

    [
      new RegExp(`^${options.prefix}rhombus-?(.+)$`),
      ([_, s]) => ({
        height: s,
        'aspect-ratio': 1,
        'clip-path': cssVarUse('rhombus-clip-path'),
      }),
      {
        autocomplete: `${options.prefix}rhombus-$size`,
      },
    ],

    [
      new RegExp(`^${options.prefix}hexagon-?(.+)$`),
      ([_, s]) => ({
        height: s,
        'aspect-ratio': '1/cos(30deg)',
        'clip-path': cssVarUse('hexagon-clip-path'),
      }),
      {
        autocomplete: `${options.prefix}hexagon-$size`,
      },
    ],

    [
      new RegExp(`^${options.prefix}octagon-?(.+)$`),
      ([_, s]) => ({
        width: s,
        'aspect-ratio': 1,
        'clip-path': cssVarUse('octagon-clip-path'),
      }),
      {
        autocomplete: `${options.prefix}octagon-$size`,
      },
    ],

    [
      new RegExp(`^${options.prefix}heart-?(.+)$`),
      ([_, s]) => ({
        height: s,
        'aspect-ratio': 1,
        '-webkit-mask-box-image': cssVarUse('heart-gradient-mask'),
        'mask-border': cssVarUse('heart-gradient-mask'),
        'clip-path': cssVarUse('heart-clip-path'),
      }),
      {
        autocomplete: `${options.prefix}heart-$size`,
      },
    ],

    [
      new RegExp(`^${options.prefix}parallelogram-?(.+)$`),
      ([_, s]) => ({
        height: s,
        'aspect-ratio': '3/2',
        'clip-path': cssVarUse('parallelogram-clip-path'),
      }),
      {
        autocomplete: `${options.prefix}parallelogram-$size`,
      },
    ],
  ]
  return rules
}
