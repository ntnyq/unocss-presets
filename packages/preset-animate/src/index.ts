import { createPreflights } from './preflight'
import { createAnimationRules, createBaseRules } from './rules'
import type { Preset } from '@unocss/core'
import type { PresetAnimateOptions } from './types'

export const presetAnimate = (options: PresetAnimateOptions = {}): Preset => {
  options.preflight = options.preflight ?? true
  options.injectMediaQuery = options.injectMediaQuery ?? true
  options.variablePrefix = options.variablePrefix ?? 'un-animate-'
  options.extendAnimations = options.extendAnimations ?? []

  return {
    name: 'unocss-preset-animate',

    rules: [...createBaseRules(options), ...createAnimationRules(options)],

    preflights: options.preflight ? createPreflights(options) : [],
  }
}

export * from './types'
export * from './animations'

export default presetAnimate
