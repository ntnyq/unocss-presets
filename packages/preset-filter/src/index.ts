import { resolveOptions } from './options'
import { createPreflights } from './preflight'
import { createFilterRules } from './rules'
import type { Preset } from '@unocss/core'
import type { PresetFilterOptions } from './types'

export const presetFilter = (options: Partial<PresetFilterOptions> = {}): Preset => {
  const resolvedOptions = resolveOptions(options)
  return {
    name: 'unocss-preset-filter',
    rules: [...createFilterRules(resolvedOptions)],
    preflights: createPreflights(resolvedOptions),
  }
}

export * from './types'
export * from './options'
export * from './filters'

export default presetFilter
