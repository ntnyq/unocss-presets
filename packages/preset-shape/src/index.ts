import { definePreset } from '@unocss/core'
import { resolveOptions } from './options'
import { createPreflights } from './preflight'
import { createShapeRules } from './rules'
import type { PresetShapeOptions } from './types'

export const presetShape = definePreset<PresetShapeOptions>((options = {}) => {
  const resolvedOptions = resolveOptions(options)
  return {
    name: 'unocss-preset-shape',
    rules: createShapeRules(resolvedOptions),
    preflights: createPreflights(resolvedOptions),
  }
})

export * from './types'
export * from './options'

export default presetShape
