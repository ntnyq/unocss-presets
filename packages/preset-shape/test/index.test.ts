import { createGenerator } from '@unocss/core'
import { describe, expect, it } from 'vitest'
import { presetShape } from '../src'

const shapes = [
  'star-100px',
  'triangle-100px',
  'rhombus-100px',
  'hexagon-100px',
  'octagon-100px',
  'heart-100px',
  'parallelogram-100px',
]

describe('default', () => {
  it('should match snapshot', async () => {
    const generator = await createGenerator({
      presets: [presetShape()],
    })
    const { css } = await generator.generate(
      shapes.map(s => `shape-${s}`).join('\n'),
    )
    expect(css).toMatchSnapshot()
  })
})

describe('custom prefix', () => {
  it('should match snapshot', async () => {
    const generator = await createGenerator({
      presets: [presetShape({ prefix: 's-' })],
    })
    const { css } = await generator.generate(
      shapes.map(s => `s-${s}`).join('\n'),
    )
    expect(css).toMatchSnapshot()
  })
})
