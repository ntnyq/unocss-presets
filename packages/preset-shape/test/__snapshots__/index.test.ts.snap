// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom prefix > should match snapshot 1`] = `
"/* layer: preflights */
:root {
--s-star-clip-path: polygon(50% 0,79% 90%,2% 35%,98% 35%,21% 90%);
--s-triangle-clip-path: polygon(50% 0,100% 100%,0 100%);
--s-rhombus-clip-path: polygon(50% 0,100% 50%,50% 100%,0 50%);
--s-hexagon-clip-path: polygon(50% -50%,100% 50%,50% 150%,0 50%);
--s-octagon-angle: calc(50% * tan(-22.5deg));
--s-octagon-clip-path: polygon(var(--s-octagon-angle) 50%, 50% var(--s-octagon-angle), calc(100% - var(--s-octagon-angle)) 50%, 50% calc(100% - var(--s-octagon-angle)));
--s-heart-gradient-mask: radial-gradient(#000 69%,#0000 70%) 84.5%/50%;
--s-heart-clip-path: polygon(-41% 0,50% 91%, 141% 0);
--s-parallelogram-skew: 30%;
--s-parallelogram-clip-path: polygon(var(--s-parallelogram-skew) 0,100% 0,calc(100% - var(--s-parallelogram-skew)) 100%,0 100%);
}
/* layer: default */
.s-star-100px{width:100px;aspect-ratio:1;clip-path:var(--s-star-clip-path);}
.s-triangle-100px{width:100px;aspect-ratio:1/cos(30deg);clip-path:var(--s-triangle-clip-path);}
.s-rhombus-100px{height:100px;aspect-ratio:1;clip-path:var(--s-rhombus-clip-path);}
.s-hexagon-100px{height:100px;aspect-ratio:1/cos(30deg);clip-path:var(--s-hexagon-clip-path);}
.s-octagon-100px{width:100px;aspect-ratio:1;clip-path:var(--s-octagon-clip-path);}
.s-heart-100px{height:100px;aspect-ratio:1;-webkit-mask-box-image:var(--s-heart-gradient-mask);mask-border:var(--s-heart-gradient-mask);clip-path:var(--s-heart-clip-path);}
.s-parallelogram-100px{height:100px;aspect-ratio:3/2;clip-path:var(--s-parallelogram-clip-path);}"
`;

exports[`default > should match snapshot 1`] = `
"/* layer: preflights */
:root {
--shape-star-clip-path: polygon(50% 0,79% 90%,2% 35%,98% 35%,21% 90%);
--shape-triangle-clip-path: polygon(50% 0,100% 100%,0 100%);
--shape-rhombus-clip-path: polygon(50% 0,100% 50%,50% 100%,0 50%);
--shape-hexagon-clip-path: polygon(50% -50%,100% 50%,50% 150%,0 50%);
--shape-octagon-angle: calc(50% * tan(-22.5deg));
--shape-octagon-clip-path: polygon(var(--shape-octagon-angle) 50%, 50% var(--shape-octagon-angle), calc(100% - var(--shape-octagon-angle)) 50%, 50% calc(100% - var(--shape-octagon-angle)));
--shape-heart-gradient-mask: radial-gradient(#000 69%,#0000 70%) 84.5%/50%;
--shape-heart-clip-path: polygon(-41% 0,50% 91%, 141% 0);
--shape-parallelogram-skew: 30%;
--shape-parallelogram-clip-path: polygon(var(--shape-parallelogram-skew) 0,100% 0,calc(100% - var(--shape-parallelogram-skew)) 100%,0 100%);
}
/* layer: default */
.shape-star-100px{width:100px;aspect-ratio:1;clip-path:var(--shape-star-clip-path);}
.shape-triangle-100px{width:100px;aspect-ratio:1/cos(30deg);clip-path:var(--shape-triangle-clip-path);}
.shape-rhombus-100px{height:100px;aspect-ratio:1;clip-path:var(--shape-rhombus-clip-path);}
.shape-hexagon-100px{height:100px;aspect-ratio:1/cos(30deg);clip-path:var(--shape-hexagon-clip-path);}
.shape-octagon-100px{width:100px;aspect-ratio:1;clip-path:var(--shape-octagon-clip-path);}
.shape-heart-100px{height:100px;aspect-ratio:1;-webkit-mask-box-image:var(--shape-heart-gradient-mask);mask-border:var(--shape-heart-gradient-mask);clip-path:var(--shape-heart-clip-path);}
.shape-parallelogram-100px{height:100px;aspect-ratio:3/2;clip-path:var(--shape-parallelogram-clip-path);}"
`;
